STAR/draft-ietf-acme-star.md

```diff
@@ -336,7 +342,9 @@ The Order resource is extended with the following attributes:
     "recurrent": true,
     "recurrent-start-date": "2016-01-01T00:00:00Z",
     "recurrent-end-date": "2017-01-01T00:00:00Z",
-    "recurrent-certificate-validity": 604800
+    "recurrent-certificate-validity": 604800,
+    "recurrent-certificate-predate": 432000,
```

Comment:
Hmmm, I thought pre-dating was a prerogative of the server, with the amount also chosen by it. Do we really need to make it an explicit parameter?

Reply:
In fact, going back to the draft, I think EKR and you both missed the relevant text that makes pre-dating a MUST:

``
To avoid the client accidentally entering a broken state,
   the "next" certificate MUST be pre-dated so that it is already valid
   when it is published at the "star-certificate" URL.
``

Given this text (which applies to each short-term cert, not just the first one), I'm not sure EKR's comment is even warranted.

Reply:
Yep, I'd missed that!

However, I think the amount of predating ultimately depends on client's use case, which the server is not necessarily aware of.

Say the client needs 6d predating (because of [Section 7.1 of Acer](https://acmccs.github.io/papers/p1407-acerA.pdf)): star-04 would force it to specify a `recurrent-certificate-validity` of at least 12d to be sure that the server predates according to its needs?  It doesn't look particularly good.

Note that in Section 4.1 we say:
```
   [...I t is RECOMMENDED that site owners pre-date their
   Web facing certificates by 5 to 7 days.
```
but we don't provide the hook for the site owner to control that.  This is an inconsistency that we should fix.

I propose we leave this client-side knob but also say that the predating is computed as max(server-side-predating, client-side-predating) to make sure that cert rotation works whatever the client says.

Reply:
I think most clients are not aware of how long the time skew they need to handle, because long skews are a rare occurrence (in my experience). So we'd better not complicate the protocol only for this reason.

Instead, we can say that the server SHOULD choose the recurrence period with this time skew in mind.

Also, if you can spend some time on an example with real dates, it would make everything much easier to understand.

Reply:
> I think most clients are not aware of how long the time skew they need to handle, because long skews are a rare occurrence (in my experience). So we'd better not complicate the protocol only for this reason.
> Instead, we can say that the server SHOULD choose the recurrence period with this time skew in mind.

I don't like the idea of baking assumptions about certs consumers's clocks in the protocol.  What is the "reference" clock skew the STAR server should account for?  It's the web one (5d)?  Or the datacenter's (0.01s)?

WRT the extra complication: it seems to me that the increase in complexity required by `recurrent-certificate-predate` rates pretty low compared to the supplement of functionality provided.
To make it completely robust we'd just need to add a default (0) and describe the algorithm to compute the "final" predating -- which depends on the refresh period.  Note that we wouldn't need the latter if I had managed to convince you and Diego to clearly separate the current and next certificate URLs, but that boat has shipped a long time ago ;-)

Reply:
> Also, if you can spend some time on an example with real dates, it would make everything much easier to understand.

Not sure I understand.  You mean in the PR itself?  Or in a comment here?

Reply:
Re: the explicit predate parameter, I'm not very happy because I think it's a nit and people will not be using it, and because it's late to add stuff to the protocol. But I can live with that.

Re: example, yes, I meant as a new section in the document. This timing thing is getting complicated!

Reply:
> Re: the explicit predate parameter, I'm not very happy because I think it's a nit and people will not be using it, and because it's late to add stuff to the protocol. But I can live with that.

I agree it's not ideal to add it now, but IMHO it's a worth addition.

> Re: example, yes, I meant as a new section in the document. This timing thing is getting complicated!

Yes, sure.  If you are OK to merge this I will add the example in a subsequent PR.

Reply:
Almost. Please add the default value, and I think we're missing a reference for the Capabilities URL.

Reply:
Also, can you please change "has undefined semantics" to "is not allowed" or something.

Reply:
> Almost. Please add the default value,

done in a1810cb

> and I think we're missing a reference for the Capabilities URL.

The question mark in the ref `{{?...}}` should do the trick automagically.

Reply:
> Also, can you please change "has undefined semantics" to "is not allowed" or something.

done in a1810cb